ext/-test-/thread/instrumentation/instrumentation.c

```diff
@@ -0,0 +1,95 @@
+#include "ruby/ruby.h"
+#include "ruby/atomic.h"
+#include "ruby/thread.h"
+#include "ruby/thread_native.h"
+
+static rb_atomic_t acquire_enter_count = 0;
+static rb_atomic_t acquire_exit_count = 0;
+static rb_atomic_t release_count = 0;
+
+void
+ex_callback(rb_event_flag_t event)
+{
+    switch(event) {
+      case RUBY_INTERNAL_EVENT_THREAD_READY:
+        RUBY_ATOMIC_INC(acquire_enter_count);
+        break;
+      case RUBY_INTERNAL_EVENT_THREAD_RESUMED:
+        RUBY_ATOMIC_INC(acquire_exit_count);
+        break;
+      case RUBY_INTERNAL_EVENT_THREAD_SUSPENDED:
+        RUBY_ATOMIC_INC(release_count);
+        break;
+    }
+}
+
+static rb_thread_hook_t * single_hook = NULL;
+
+static VALUE
+thread_counters(VALUE thread)
+{
+    VALUE array = rb_ary_new2(3);
+    rb_ary_push(array, UINT2NUM(acquire_enter_count));
+    rb_ary_push(array, UINT2NUM(acquire_exit_count));
+    rb_ary_push(array, UINT2NUM(release_count));
+    return array;
+}
+
+static VALUE
+thread_reset_counters(VALUE thread)
+{
+    RUBY_ATOMIC_SET(acquire_enter_count, 0);
+    RUBY_ATOMIC_SET(acquire_exit_count, 0);
+    RUBY_ATOMIC_SET(release_count, 0);
+    return Qtrue;
+}
+
+static VALUE
+thread_register_callback(VALUE thread)
+{
+    single_hook = rb_thread_internal_event_new(
+        *ex_callback,
+        RUBY_INTERNAL_EVENT_THREAD_READY | RUBY_INTERNAL_EVENT_THREAD_RESUMED | RUBY_INTERNAL_EVENT_THREAD_SUSPENDED
+    );
+
+    return Qnil;
+}
+
+static VALUE
+thread_unregister_callback(VALUE thread)
+{
+    if (single_hook) {
+        rb_thread_internal_event_delete(single_hook);
+        single_hook = NULL;
+    }
+
+    return Qnil;
+}
+
+static VALUE
+thread_register_and_unregister_callback(VALUE thread)
+{
+    rb_thread_hook_t * hooks[5];
+    for (int i = 0; i < 5; i++) {
+        hooks[i] = rb_thread_internal_event_new(*ex_callback, RUBY_INTERNAL_EVENT_THREAD_READY);
+    }
+
+    if (!rb_thread_internal_event_delete(hooks[4])) return Qfalse;
+    if (!rb_thread_internal_event_delete(hooks[0])) return Qfalse;
+    if (!rb_thread_internal_event_delete(hooks[3])) return Qfalse;
+    if (!rb_thread_internal_event_delete(hooks[2])) return Qfalse;
+    if (!rb_thread_internal_event_delete(hooks[1])) return Qfalse;
```

Comment:
Why 4→0→3→2→1 order?

Reply:
To test both removing the last in the list, the first and in the middle. But it's mostly random.